sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```
@@ -4701,4 +4987,425 @@ static UUID randomUuid(long msb, long lsb) {
+    }
+
+    private static class NonTransientPointOperationResult {
+        private final ResourceResponse<Document> response;
+        private final CosmosException exception;
+
+        public NonTransientPointOperationResult(CosmosException exception) {
+            checkNotNull(exception, "Argument 'exception' must not be null.");
+            this.exception = exception;
+            this.response = null;
+        }
+
+        public NonTransientPointOperationResult(ResourceResponse<Document> response) {
+            checkNotNull(response, "Argument 'response' must not be null.");
+            this.exception = null;
+            this.response = response;
+        }
+
+        public boolean isError() {
+            return this.exception != null;
+        }
+
+        public CosmosException getException() {
+            return this.exception;
+        }
+
+        public ResourceResponse<Document> getResponse() {
+            return this.response;
+        }
+    }
+
+    private static class ScopedDiagnosticsFactory implements DiagnosticsClientContext {
+
+        private AtomicBoolean isMerged = new AtomicBoolean(false);
+        private final DiagnosticsClientContext inner;
+        private final ConcurrentLinkedQueue<CosmosDiagnostics> createdDiagnostics;
+
+        public ScopedDiagnosticsFactory(DiagnosticsClientContext inner) {
+            checkNotNull(inner, "Argument 'inner' must not be null.");
+            this.inner = inner;
+            this.createdDiagnostics = new ConcurrentLinkedQueue<>();
+        }
+
+        @Override
+        public DiagnosticsClientConfig getConfig() {
+            return inner.getConfig();
+        }
+
+        @Override
+        public CosmosDiagnostics createDiagnostics() {
+            assert !this.isMerged.get() : "Not allowed to create new diagnostics after merging.";
```

should we throw an exception here or just return null and log the error?

could there be a such race condition happens that:

- we get response from one region,
-  diagnostics merged
- even though it tried to cancel the request to the other regions, but still we reached here for the other regions

---

Yes - I added the assert to validate the assumptions during testing and forgot to revert it.

---

Fixed